webapp/packages/plugin-codemirror6/src/ReactCodemirror.tsx

```diff
@@ -0,0 +1,128 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2023 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+
+import { EditorView } from 'codemirror6';
+import { forwardRef, useEffect, useImperativeHandle, useLayoutEffect, useState } from 'react';
+
+import { EditorState, Annotation, StateEffect } from '@codemirror/state';
+import type { ViewUpdate } from '@codemirror/view';
+
+import type { IEditorRef } from './IEditorRef';
+import type { IReactCodeMirrorProps } from './IReactCodemirrorProps';
+
+const External = Annotation.define<boolean>();
+
+export const ReactCodemirror = forwardRef<IEditorRef, IReactCodeMirrorProps>(function ReactCodemirror({
+  value,
+  extensions,
+  readonly,
+  editable,
+  autoFocus,
+  onChange,
+  onUpdate,
+}, ref) {
+  const [container, setContainer] = useState<HTMLDivElement | null>(null);
+  const [view, setView] = useState<EditorView | null>(null);
+  const [state, setState] = useState<EditorState | null>(null);
+
+  const defaultTheme = EditorView.theme({
+    '&': {
+      width: '100%',
+      height: '100%',
+    },
+  });
+
+  const ext = [defaultTheme];
+
+  if (editable === false) {
+    ext.push(EditorView.editable.of(false));
+  }
+
+  if (readonly) {
+    ext.push(EditorState.readOnly.of(true));
+  }
+
+  if (extensions) {
+    ext.push(extensions);
+  }
+
+  if (onChange || onUpdate) {
+    const updateListener = EditorView.updateListener.of((update: ViewUpdate) => {
+      const remote = update.transactions.some(tr => tr.annotation(External));
+
+      if (update.docChanged && !remote) {
+        const doc = update.state.doc;
+        const value = doc.toString();
+
+        onChange?.(value, update);
+      }
+
+      onUpdate?.(update);
+    });
+
+    ext.push(updateListener);
+  }
+
+  useEffect(() => {
+    if (container) {
+      const es = EditorState.create({
+        doc: value,
+        extensions: ext,
+      });
```

Comment:
seems like it can be set only once in the useState
```tsx
const [editorState] = useState(() => EditorState.create({
        doc: value,
        extensions: ext,
      }));
```

Reply:
you can also skip `extensions` and `doc` for init, because you have logic later to update them